[WIP] Porting ros2

hsd-dev opened the pull request:
Originally from https://github.com/PepperlFuchs/pf_lidar_ros_driver/pull/82 started by @wawanbreton

- [x] CI: https://github.com/PepperlFuchs/pf_lidar_ros_driver/pull/83/commits/de9e3adfc232b766271b7c5a3e57a3477c90528d, https://github.com/PepperlFuchs/pf_lidar_ros_driver/pull/83/commits/7d354f603cf16a2f3389d20c122cb16a3f41dd2c
- [x] README: https://github.com/PepperlFuchs/pf_lidar_ros_driver/pull/83/commits/8d3a2f918bda80feaf5ec71001f82735d4efd633
- [x] pf_interfaces
  - [x] package.xml and CMakeLists.txt
- [x] pf_description
  - [x] package.xml and ~~CMakeLists.txt~~ setup.py: https://github.com/PepperlFuchs/pf_lidar_ros_driver/pull/83/commits/9476ad63b870801110c965d7b480135093a04f58
  - [x] launch files: https://github.com/PepperlFuchs/pf_lidar_ros_driver/pull/83/commits/64c4d08a72eaa3a438f659ed5c97431c0ac8c611
  - [x] update rviz config: https://github.com/PepperlFuchs/pf_lidar_ros_driver/pull/83/commits/d01c6d13dc8092307f9531238643dd0afe73d084
- [X] pf_driver
  - [x] package.xml and CMakeLists.txt: https://github.com/PepperlFuchs/pf_lidar_ros_driver/pull/83/commits/0248e7493d6562fe3ebc6dc52c860437c7be5634
  - [x] Node, logging, interfaces, callbacks: https://github.com/PepperlFuchs/pf_lidar_ros_driver/pull/83/commits/0248e7493d6562fe3ebc6dc52c860437c7be5634 
  - [x] launch files
  - [x] tests
  - [x] cpp code refactoring to improve parameter management: https://github.com/PepperlFuchs/pf_lidar_ros_driver/commit/1244f3d42f1cd3584cb84fd5127f365a69f4f52f (currently removed since https://github.com/PepperlFuchs/pf_lidar_ros_driver/pull/77 and https://github.com/PepperlFuchs/pf_lidar_ros_driver/pull/81 changes quite some cpp code)

hsd-dev committed:
rename publisher classes since R2300 could also be single layered

hsd-dev committed:
parameterize number of layers since it could vary for future versions

hsd-dev committed:
data publisher differs for R2300 based on device_family parameter

hsd-dev committed:
apply clang-tidy formatting

hsd-dev committed:
re-arrange functions

hsd-dev committed:
fix device_family incorrect

hsd-dev committed:
get device part number

Currently correction params are defined for R2300 based on their part numbers.
Need to verify if that is the intended way.

hsd-dev committed:
define correction params for the available device

hsd-dev committed:
[WIP] apply correction function to the pointcloud

The corrected pointcloud is wrong. Mostly issues with units

hsd-dev committed:
apply clang-tidy formatting

hsd-dev committed:
correct implementation of the correction functions

hsd-dev committed:
no need to transform individual clouds since apply_correction does the job

hsd-dev committed:
do not publish laserscan data for PointcloudPublisher

this does not make sense especially after apply correction function

hsd-dev committed:
custom project laser function including correction calculations

hsd-dev committed:
map quadratic equation coefficients with inclination angle

hsd-dev committed:
close reader and writer only on terminate to avoid getting called twice

hsd-dev committed:
detect and handle connection failure

hsd-dev committed:
add async read with timeout to avoid blocking calls

hsd-dev committed:
recognize device based on device_family

hsd-dev committed:
clang tidy

Commit:
Basically working version of ROS2 porting.
Not all parameters are implemented, and some deadlocks occur.

Commit:
Cleaning

hsd-dev committed:
print timeout message only if success

hsd-dev committed:
basic porting of description package - builds

hsd-dev committed:
port bringup launch

hsd-dev committed:
bringup rviz

current rviz config is not supported in ROS2

wawanbreton commented:
I got permission denied when pushing to this branch, can you give me some rights ?

ipa-hsd commented on pf_driver/include/pf_driver/pf/r2000/pfsdp_2000.hpp:
```diff
@@ -28,119 +28,46 @@ class PFSDP_2000 : public PFSDPBase
-    else if (level == 23)
-    {
-      config_->max_num_points_scan = config.max_num_points_scan;
+    for(const auto &parameter : parameters)
```
nice :)

wawanbreton replied:
Dynamic parameters are indeed much simpler in ROS2 !

hsd-dev committed:
update CI for ROS2

hsd-dev committed:
fix clang tidy

ipa-hsd commented:
The current code seems to be valid only for `foxy`. We might need to have separate branches for `galactic` and upwards https://github.com/PepperlFuchs/pf_lidar_ros_driver/runs/6810174179?check_suite_focus=true#step:4:685

wawanbreton commented:
> The current code seems to be valid only for `foxy`. We might need to have separate branch for `galactic` and upwards 

What makes you say that ? I'm working with galaxy only

ipa-hsd commented:
@wawanbreton please check the CI output linked above. Didn't you get that error?

wawanbreton commented:
No I didn't get the error, pcl-conversion is actually available for galactic

ipa-hsd commented:
right, we have commented out the deps https://github.com/PepperlFuchs/pf_lidar_ros_driver/blob/porting-ros2/pf_driver/package.xml#L13. Since CI starts with a clean docker, these packages have not been installed.

hsd-dev committed:
add deps

hsd-dev committed:
remove humble from CI

hsd-dev committed:
add dep to interfaces

ipa-hsd commented:
OK the CI seems to be working now: https://github.com/PepperlFuchs/pf_lidar_ros_driver/runs/6811233956?check_suite_focus=true#step:4:1582

hsd-dev committed:
update build and usage instructions and CI badge

wawanbreton commented:
My commit containing the DAE files instead of the STL has been wiped out, was it intentional ?

ipa-hsd commented:
Also, take a look at the CI output linked above
```
/root/target_ws/src/pf_lidar_ros_driver/pf_driver/src/pf/pf_interface.cpp:220:25: error: missing template arguments before ‘feed_time’
    220 |   std::chrono::duration feed_time = std::min(duration, std::chrono::milliseconds(std::chrono::seconds(60)));
        |                         ^~~~~~~~~
  /root/target_ws/src/pf_lidar_ros_driver/pf_driver/src/pf/pf_interface.cpp:221:46: error: ‘feed_time’ was not declared in this scope
    221 |   watchdog_timer_ = node_->create_wall_timer(feed_time, std::bind(&PFInterface::feed_watchdog, this));
```

wawanbreton commented:
Commit restored. Please don't force-push unless you are alone on a repos :)

Commit:
Removed missing install directory

Commit:
Fixed pf_interface build on foxy

wawanbreton commented:
Build error should be fixed, I used a declaration that is only available with C++17 but foxy targets C++14

ipa-hsd commented:
Should we remove these two commits for now? https://github.com/PepperlFuchs/pf_lidar_ros_driver/pull/83/commits/3f1385c7fd3d02e49c5014705406aa031fbfd073 and https://github.com/PepperlFuchs/pf_lidar_ros_driver/pull/83/commits/c6c290fb428674dd6ca729039dc04990b9546594. For this PR I want to focus on porting existing functionality to ROS2 and not add new features mid-way. I can create a new PR with those 2 commits stacks behind this one.

wawanbreton commented:
I completely agree, I was wondering about splitting them to an other PR. Maybe this should be the case for https://github.com/PepperlFuchs/pf_lidar_ros_driver/pull/83/commits/d75100945368923990de207e8ce2ad479ae97bdf too ?

hsd-dev committed:
fix get_parameter device family

wawanbreton commented:
Hi @ipa-hsd, I'm starting to rework the parameters stack because some things are not properly done at initialization. I'm wondering about the parameters philosophy, I can see two possible options :

1. All parameters given to driver are set to the LIDAR, other parameters are forced to default values. LIDAR previous configuration is not taken care of.
2. All parameters given to driver are set to the LIDAR, other parameters are read from the LIDAR and stored into actual parameter values.

Option 1 is more repeatable because whatever you did before, the LIDAR will always start in the same configuration. Option 2 considers you may want to store some configuration into the LIDAR definitely. Personnally, I prefer option 1 because we can easily set all the required parameters into a config file, so I don't really see a need to store configuration on long-term. What do you think should be implemented ?

ipa-hsd commented:
Thanks for taking a look into this. 

>  some things are not properly done at initialization

Anything specific?

Maybe @ptruka has something to say about this based on the customer feedback? 

@wawanbreton also I am starting to think about the fact the base cpp code (not ROS specific parts) is starting to diverge from the code for ROS1. I am not sure to what extent it is possible to reuse the same classes, but something we might want to keep in mind. One possibility could be whatever inconsistencies / bugs you have fixed in this PR, we could propagate it back to the ROS1 code as well.

wawanbreton commented:
> Anything specific?

Yesterday I implemented start_angle parameter, but then figured out that it was not applied at startup. I'm working on it.

> One possibility could be whatever inconsistencies / bugs you have fixed in this PR, we could propagate it back to the ROS1 code as well

Until now, I only fixed the bugs I created by changing the parameters description. The management is very different from the one with ROS1, so I'm not sure we could have the same code base for both. I'd be happy to really refactor this part, but I would understand you want to keep consistency between versions. However, I only changed the ROS-related code but all low-level communication code is still the same.

ipa-hsd commented:
As long as we ensure that the behaviour is the same in ROS1 and ROS2, it should be fine.

wawanbreton commented:
> As long as we ensure that the behaviour is the same in ROS1 and ROS2, it should be fine. I can think of some tests for that.

Great, I'll keep going this way. And It would make me feel better to have some tests indeed :)

hsd-dev committed:
apply correction as ROS param

ipa-hsd commented on pf_driver/src/pf/pfsdp_protocol.cpp:
```diff
@@ -204,6 +203,12 @@ void PFSDPBase::declare_critical_parameters()
+  descriptorTransport.name = "IP transport port";
+  descriptorTransport.description = "Can be either tcp or udp";
+  descriptorTransport.read_only = true;
+  node_->declare_parameter<std::string>("transport", "tcp", descriptorTransport);
```
Also I didn't notice this earlier, but it's better to separate the protocol-specific code from ROS-specific code. The protocol-specific code is intended to be used like a pure cpp library with no ROS dependency, apart from logging.

wawanbreton commented:
> it would make sense to use this PR to make only ROS-specific changes and make a new PR for cpp code improvements.

OK, I didn't understand this. The thing is, proper management of parameters will require some architectural changes, so I thought it would be the right time. But I can do something a bit less intrusive yet, even if it is a bit dirty, and make it clean later, if that's what you want.

> The protocol-specific code is intended to be used like a pure cpp library with no ROS dependency, apart from logging

Noted, it perfectly makes sense !

ipa-hsd commented:
> The thing is, proper management of parameters will require some architectural changes, so I thought it would be the right time.

I agree, but some logic has changed and couple of new features have been introduced in those PRs. I have added that commit as part of TODO in this PR's description.

wawanbreton commented:
Ok, I will wait for your go to work on this.

ipa-hsd commented:
The CI is failing because the code is not formatted properly https://github.com/PepperlFuchs/pf_lidar_ros_driver/runs/7018472659?check_suite_focus=true#step:4:174
Could you apply the clang formatting? (If you are using VS Code, right click on the file and "Format document with..."). 

The link above shows which documents need to be updated.

hsd-dev committed:
fix product string

wawanbreton commented:
I'm not using VS Code, but I tried applying formatting using clang-format. However, the whole file is changed and becomes very different... Is there a formatting profile you are implicitely using ?

ipa-hsd commented:
This one: https://github.com/PepperlFuchs/pf_lidar_ros_driver/blob/main/.clang-format

wawanbreton commented:
Seems I got it right :) I configured clang-format to style "FIle" so it probably found it by itself. Thank you for pointing it to me, I'd like to use it for my own projects !

hsd-dev committed:
projectLaser from laser_geometry already calculates x,y

hsd-dev committed:
changed coefficients to radians

hsd-dev committed:
combine calculations for correction on and off

hsd-dev committed:
publish laser scans and transform when no correction

hsd-dev committed:
call project_laser when correction is on

ptruka committed:
Merge pull request #77 from PepperlFuchs/correction_func

Apply correction function for the pointcloud (R2300)
Fixed Bug that caused the driver to not start correctly (R2000/R2300)

hsd-dev committed:
remove std outs

hsd-dev committed:
read_result always exists

hsd-dev committed:
release handle at init and at shutdown

hsd-dev committed:
implement reconnectivity with conditional variable

hsd-dev committed:
reduce timeout for no packets received

hsd-dev committed:
fix clang

hsd-dev committed:
move function to cpp file

hsd-dev committed:
implement async read for TCP

hsd-dev committed:
handle TCP connect failure

hsd-dev committed:
let device provide the port number always

hsd-dev committed:
fix clang

hsd-dev committed:
to enable intensity information

ptruka committed:
Merge pull request #81 from PepperlFuchs/reconnectivity

added reconnectivity feature

hsd-dev committed:
causes publish freq to drop at higher scan frequencies and max samples per scan

not sure why this sleep was added

ipa-hsd commented:
@wawanbreton both https://github.com/PepperlFuchs/pf_lidar_ros_driver/pull/77 and https://github.com/PepperlFuchs/pf_lidar_ros_driver/pull/81 have been merged now. Could you apply https://github.com/PepperlFuchs/pf_lidar_ros_driver/commit/1244f3d42f1cd3584cb84fd5127f365a69f4f52f to the latest `main` branch? Then we can rebase this PR.

ptruka committed:
Merge pull request #85 from PepperlFuchs/fix_uhd_perf

Fix UHD performance issue at frequencies higher than 30Hz

ipa-hsd commented:
No worries, I can make those changes based on https://github.com/PepperlFuchs/pf_lidar_ros_driver/commit/1244f3d42f1cd3584cb84fd5127f365a69f4f52f and you can carry on from there.

hsd-dev committed:
remove mutex

dynamically changing parameters causes deadlock

Commit:
Moved sources to src folder

Commit:
Splitted headers to have a single class by file

Commit:
Splitted existing cpp files to match headers

Commit:
Added cpp files for each header when possible

Commit:
Code cleaning and minor optimizations

Commit:
Applied clang formatting

Commit:
Fixed missing include

hsd-dev committed:
Merge branch 'main' into fix_deadlock

MichaelGrupp committed:
Allow to disable RViz & pf_bringup in launch file.

This allows better integration in headless systems and robots that
already have a robot_state_publisher.

MichaelGrupp committed:
Allow to set hmi_application_bitmap

ptruka committed:
Merge pull request #86 from PepperlFuchs/fix_deadlock

fix deadlock

hsd-dev committed:
remove references to mutex after refactoring

hsd-dev committed:
header only has function declaration

hsd-dev committed:
apply clang formatting

hsd-dev committed:
utilize port specified in config for UDP connection

hsd-dev committed:
Basically working version of ROS2 porting.
Not all parameters are implemented, and some deadlocks occur.

# Conflicts:
#	pf_driver/include/pf_driver/pf/pf_interface.h
#	pf_driver/include/pf_driver/pf/pf_packet.h
#	pf_driver/include/pf_driver/pf/pfsdp_protocol.hpp
#	pf_driver/include/pf_driver/pf/pipeline.h
#	pf_driver/include/pf_driver/pf/r2000/pfsdp_2000.hpp
#	pf_driver/include/pf_driver/pf/r2300/pfsdp_2300.hpp
#	pf_driver/include/pf_driver/ros/scan_publisher.h
#	pf_driver/msg/PFHeader.msg
#	pf_driver/msg/PFR2000Header.msg
#	pf_driver/msg/PFR2300Header.msg
#	pf_driver/src/pf/pf_packet.cpp
#	pf_interfaces/msg/PFHeader.msg
#	pf_interfaces/msg/PFR2000Header.msg
#	pf_interfaces/msg/PFR2300Header.msg
#	src/pf_driver/CMakeLists.txt
#	src/pf_driver/include/pf_driver/pf/pf_parser.h
#	src/pf_driver/msg/PFHeader.msg
#	src/pf_driver/msg/PFR2000Header.msg
#	src/pf_driver/msg/PFR2300Header.msg
#	src/pf_driver/src/pf/pf_interface.cpp
#	src/pf_driver/src/pf/pfsdp_protocol.cpp
#	src/pf_driver/src/ros/pf_data_publisher.cpp
#	src/pf_driver/src/ros/ros_main.cpp
#	src/pf_driver/tests/pfsdp_protocol.cpp

hsd-dev committed:
move interfaces package to src folder

ipa-hsd commented:
@wawanbreton I have rebased this PR from https://github.com/PepperlFuchs/pf_lidar_ros_driver/pull/87. Now we are good to go.

hsd-dev committed:
apply clang

hsd-dev committed:
create new package for interfaces

hsd-dev committed:
port header message serialization

hsd-dev committed:
apply clang

ptruka committed:
Merge pull request #87 from wawanbreton/code_refactoring

Code refactoring

hsd-dev committed:
apply clang

hsd-dev committed:
remove references to dynamic reconfigure

hsd-dev committed:
update message signatures and headers

hsd-dev committed:
Merge pull request #99 from PepperlFuchs/fix_udp_port

Fix udp port

hsd-dev committed:
port nodes and interfaces to ROS2

hsd-dev committed:
update message signatures

hsd-dev committed:
fix rebase and update ROS 2 interfaces

hsd-dev committed:
fix c++ includes

hsd-dev committed:
add pf_interfaces to CMakeLists

hsd-dev committed:
apply clang

hsd-dev committed:
apply clang

hsd-dev committed:
Merge pull request #93 from MichaelGrupp/launch-args

Allow to disable RViz & pf_bringup in launch file.

hsd-dev committed:
Merge branch 'main' into application_bitmap

hsd-dev committed:
Merge pull request #94 from MichaelGrupp/application_bitmap

Allow to set hmi_application_bitmap with dynamic_reconfigure

hsd-dev committed:
port is optional

hsd-dev committed:
default packet type is C

hsd-dev committed:
add config files for R2300 device family 7 (single layer)

hsd-dev committed:
Merge branch 'main' into r2300_single_layer

ptruka committed:
Merge pull request #101 from PepperlFuchs/r2300_single_layer

R2300 single layer

ptruka committed:
Update r2300.launch

fixed typo

hsd-dev committed:
replace dynamic reconf params with ROS 2 params

hsd-dev committed:
use yaml-cpp to read nested params from yaml file

ROS 2 currently does not support mixed typed nested params

hsd-dev committed:
fix clang

hsd-dev committed:
fix clang

hsd-dev committed:
add empty default value to parameter to support foxy

ValerioMagnago committed:
Move stamping close to data reading.

Before this commit the scan message was stamped just before publishing the
data. In the LaserScan message of ros the timestamp in the header is
the acquisition time of the first ray in the scan.

Commit:
Fixes type of port parameter in cfg files

The port parameter was declared as an int_t in the cfg files but in the
code it is expected to be a string. When the dynamic_reconfigure server
is started, the data type on the parameter server was changed from
string to int. This resulted in getParam of the release 1.2.0 to return
False effectively killing any respawn functionality.

ValerioMagnago committed:
Clean termination of ros_main node.

Closes: PepperlFuchs/pf_lidar_ros_driver#105

hsd-dev committed:
apply clang

hsd-dev committed:
safe feed time within the defined timeout

ptruka committed:
Merge pull request #109 from PepperlFuchs/safe_feed_time

safe feed time within the defined timeout

ptruka committed:
Merge pull request #104 from ValerioMa/main

Move stamping closer to data reading.

ptruka committed:
Merge pull request #106 from kalectro/main

Fixes type of port parameter in cfg files

ptruka committed:
Merge pull request #108 from ValerioMa/terminate_cleanly

Clean rosnode termination.

hsd-dev committed:
always chooses v6 for some reason. explicitly forcing to choose v4

get an error: bind: Address family not supported by protocol

hsd-dev committed:
declare params of different types

hsd-dev committed:
de-serialize buffer ROS 2 style

hsd-dev committed:
convert sec to milisec

hsd-dev committed:
fix rebasing issues

hsd-dev committed:
use memcpy to parse buffer to header message

hsd-dev committed:
port time now to ros 2

hsd-dev committed:
port launch files

hsd-dev committed:
fix clang

hsd-dev committed:
create var for project sources

hsd-dev committed:
add empty gtest

hsd-dev committed:
add test to parse R2000 packet of type C - test fails currently

hsd-dev committed:
fix clang

hsd-dev committed:
remove linting tests

hsd-dev committed:
update parser test to read dump

hsd-dev committed:
fix clang

hsd-dev committed:
add gtest to test pipeline by reading dump

needs an external running TCP server; fix it

hsd-dev committed:
update test to keep reader in scope

hsd-dev committed:
dump more data

hsd-dev committed:
create tests folder in include

hsd-dev committed:
use RCLCPP_INFO instead of cout

hsd-dev committed:
add tcp server to the test

hsd-dev committed:
memcpy fields individually

nested struct is causing trouble

hsd-dev committed:
causes bad_weak_ptr

hsd-dev committed:
removed debug print

hsd-dev committed:
memcpy r2300 header fields individually

hsd-dev committed:
dict way of declaring params not working

hsd-dev committed:
fix params yaml ros 2 style

hsd-dev committed:
fix clang

hsd-dev committed:
use RCPCPP logging instead of std::cout

hsd-dev committed:
spin node

timers won't work without it

hsd-dev committed:
update rviz config

hsd-dev committed:
use global node for PF data publishers

hsd-dev committed:
fix clang

hsd-dev committed:
viz scanner model in rviz

hsd-dev committed:
include xacro in r2300 launch

ipa-hsd commented:
@chenwu-cc thanks for testing the branch. I thought the problem was only on my PC. I will test it today again and get back to you.

chenwu-cc commented:
> 

May I ask if I use the galactic version could I run porting ros2.

hsd-dev committed:
calculate scan params after init

ipa-hsd commented:
@chenwu-cc could you please try this PR again? I fixed the issue.

hsd-dev committed:
fix readme

hsd-dev committed:
fix readme

hsd-dev committed:
update r2300 rviz config

hsd-dev committed:
fix r2300 launch and yaml

hsd-dev committed:
fix watchdogtimeout value for r2300

hsd-dev committed:
fix reconnect crash

hsd-dev committed:
start watchdog timer only if the device is R2000

hsd-dev committed:
add missing params for R2300

hsd-dev committed:
fix param set failure

hsd-dev committed:
add missing params

hsd-dev committed:
foxy is already eol

hsd-dev committed:
TimePanel display is not available in Foxy

hsd-dev committed:
fix get product string

hsd-dev committed:
add single layer launch file

hsd-dev committed:
implement missing params

hsd-dev committed:
scan_frequency is already declared in pfsdp_base

hsd-dev committed:
params defined in ros_main

hsd-dev committed:
fix double declaration of params

hsd-dev committed:
causes reclaration of param error

based on the device type PFSDP base is reinitialized

hsd-dev committed:
set parameter set callback in derived classes

hsd-dev committed:
fix parameters wih on/off values

hsd-dev committed:
fkix formatting

hsd-dev committed:
undefined behaviour when destroying any derived packet

refer: https://github.com/PepperlFuchs/pf_lidar_ros_driver/issues/118

hsd-dev committed:
fix ParameterAlreadyDeclaredException during reconnectivity

hsd-dev committed:
add nanoseconds value -- only warning for foxy / galactic, but error for humble

hsd-dev committed:
fix error with finding jsoncpp lib

issue: https://github.com/open-source-parsers/jsoncpp/issues/1356
example fix: https://github.com/tesseract-robotics/trajopt/pull/299

hsd-dev committed:
add humble and iron to industrial_ci

hsd-dev committed:
samples_per_scan is of type int

hsd-dev committed:
return success of set_parameter

hsd-dev committed:
fix more parameters with on/off values

hsd-dev committed:
fix clang

hsd-dev committed:
fixes for r2300 single layer

hsd-dev committed:
rviz config is defined in description package

hsd-dev committed:
add .gitignore

hsd-dev committed:
fix rviz config for single layer

hsd-dev committed:
fix again rviz config for single layer

hsd-dev committed:
handle sigint to stop transmission and release handle

hsd-dev committed:
change build to cmake

hsd-dev committed:
driver has been tested for humble

hsd-dev committed:
update tests to launch_testing

hsd-dev committed:
fix launch_testing dep

hsd-dev committed:
add missing deps
